src/cms/views/utils/search_content_ajax.py

```diff
@@ -0,0 +1,118 @@
+import re
+import logging
+import json
+
+from django.contrib.auth.decorators import login_required
+from django.http import JsonResponse
+from django.views.decorators.http import require_POST
+
+from backend.settings import WEBAPP_URL
+from ...decorators import region_permission_required
+from ...models.regions.region import Region
+
+logger = logging.getLogger(__name__)
+
+# The maximum number of results returned by `search_content_ajax`
+MAX_RESULT_COUNT = 20
+
+
+def format_object_translation(object_translation, typ):
+    """
+    Formats the [poi/event/page]-translation as json
+
+    :param object_translation: A translation object which has a title and a permalink
+    :type object_translation: ~cms.models.events.event.Event or ~cms.models.pages.page.Page or ~cms.models.pois.poi.POI
+    :param typ: The type of this object
+    :type typ: str
+    :return: A dictionary with the title, url and type of the translation object
+    :rtype: dict
+    """
+    return {
+        "title": object_translation.title,
+        "url": f"{WEBAPP_URL}/{object_translation.permalink}",
+        "type": typ,
+    }
+
+
+def find_objects(object_manager, query, language_slug):
+    """
+    Filters all object translations from ``object_manager`` of this language
+    that match ``query`` and returns them and their ranking
+
+    :param object_manager: The manager for a specific model
+    :type object_manager: ~django.db.models.Manager
+    :param query: The query string used for filtering the objects
+    :type query: str
+    :param language_slug: The language slug
+    :type language_slug: str
+    :return: an iterator over all objects that match query and their search priority
+    :rtype: Iterator[:class:`tuple`]
+    """
+    for obj in object_manager.all():
+        if obj.archived:
+            continue
+        object_translation = obj.get_public_translation(language_slug)
+        if not object_translation:
+            continue
+
+        title = object_translation.title
+        slug = object_translation.slug
+
+        if re.search(query, title, re.IGNORECASE):
+            yield (2, object_translation)
+        elif re.search(query, slug, re.IGNORECASE):
+            yield (2, object_translation)
```

Comment:
```suggestion
def find_objects(manager, obj_type, region, language_slug, query, hierarchical=False):
    """
    Filters all object translations from ``manager`` of this language that match the ``query``

    :param manager: The manager for a specific model
    :type manager: ~django.db.models.Manager

    :param obj_type: The content type of the queried translations
    :type obj_type: str

    :param region: The current region
    :type region: ~cms.models.regions.region.Region

    :param language_slug: The language slug
    :type language_slug: str

    :param query: The query string used for filtering the objects
    :type query: str

    :param hierarchical: Whether or not the given model is hiearchical (which impacts the method for filtering out
                         archived results)
    :type hierarchical: bool

    :return: A list of translation objects (formatted as dict)
    :rtype: list [ dict ]
    """
    archived_flag = "explicitly_archived" if hierarchical else "archived"
    result = (
        manager.objects.filter(
            **{
                obj_type + "__region": region,
                obj_type + "__" + archived_flag: False,
                "language__slug": language_slug,
                "status": status.PUBLIC,
            }
        )
        .filter(Q(slug__icontains=query) | Q(title__icontains=query))
        .distinct(obj_type)
    )
    # If the object type is hierarchical, filter out implicitly archived objects (objects with archived ancestors)
    if hierarchical:
        result = filter(lambda x: not getattr(x, obj_type).implicitly_archived, result)
    return [format_object_translation(obj, obj_type) for obj in result]
```
Ok, so this would be my approach to do the search on the database level instead of doing a regex search in python.
Especially with many entries in the database, this should give a significant performance improvement, although I didn't do any benchmarking on this.